### PR TITLE
fix 'Assertion `px != 0' failed.' when initializing

### PR DESCRIPTION
**Pull request by christian-rauch**
This PR fixes issue https://github.com/ros-planning/warehouse_ros_mongo/issues/14.

Connecting to the data base causes the exeption:
```rviz: /usr/include/boost/smart_ptr/scoped_ptr.hpp:105: T* boost::scoped_ptr<T>::operator->() const [with T = mongo::AtomicWord<unsigned int>]: Assertion `px != 0' failed.```
with newer versions of the mongo DB client libraries on Ubuntu 18.04.

The fix is based on information from bug report [CXX-561](https://jira.mongodb.org/browse/CXX-561?focusedCommentId=856529&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-856529) in the official C++ driver bug tracker:

> You must initialize the driver to ensure that Mongo initializer blocks are executed. You can do this by calling mongo::client::initialize or stack allocating a mongo::client::GlobalInstance in the main() block of your program.

### CODE REVIEW

File: test/test_mongo_helpers.h [hunk 1]
```diff
@@ -50,20 +51,20 @@ inline bool operator==(const Pose& p1, const Pose& p2)
 
 }
 
-inline MongoMetadata& downcastMetadata(Metadata::ConstPtr metadata) const {
-  return *(const_cast<MongoMetadata*>(static_cast<const MongoMetadata*>(metadata.get())));
+inline warehouse_ros_mongo::MongoMetadata& downcastMetadata(warehouse_ros::Metadata::ConstPtr metadata) {
```
Comment (rhaschke) [inline test/test_mongo_helpers.h]: Please revert these additional namespace prefixes.
Why did you change constness of the method? If there is no particular reason, please revert this as well.

File: test/test_mongo_helpers.h [hunk 2]
```diff
@@ -50,20 +51,20 @@ inline bool operator==(const Pose& p1, const Pose& p2)
...
+  return *(const_cast<warehouse_ros_mongo::MongoMetadata*>(static_cast<const warehouse_ros_mongo::MongoMetadata*>(metadata.get())));
 }
 
-inline MongoQuery& downcastQuery(Query::ConstPtr query) const {
-  return *(const_cast<MongoQuery*>(static_cast<const MongoQuery*>(query.get())));
+inline warehouse_ros_mongo::MongoQuery& downcastQuery(warehouse_ros::Query::ConstPtr query) {
```
Comment (rhaschke) [inline test/test_mongo_helpers.h]: Same here.

**Comment by christian-rauch** [comment]
This PR now also fixes the test compilation errors in issue https://github.com/ros-planning/warehouse_ros_mongo/issues/7.
Since the ROS master is not running in the CI instance, the tests will wait indefinitely for a connection. I, therefore, disabled the tests for now.

File: CMakeLists.txt [hunk 3]
```diff
@@ -54,10 +54,11 @@ set(warehouse_srcs
 add_library(warehouse_ros_mongo SHARED ${warehouse_srcs})
 target_link_libraries(warehouse_ros_mongo ${catkin_LIBRARIES} ${MongoDB_LIBRARIES} ${OPENSSL_LIBRARIES} ${Boost_LIBRARIES})
 
-if(CATKIN_ENABLE_TESTING)
-  catkin_add_gtest(test_warehouse_ros_mongo_cpp test/test_warehouse_ros_mongo.cpp)
-  target_link_libraries(test_warehouse_ros_mongo_cpp warehouse_ros_mongo)
-endif()
+# TODO: start ROS master in CI, skip tests for now
```
Comment (rhaschke) [inline CMakeLists.txt]: Can you please provide a rostest file to start the mongodb server?
I think that's a rather easy task and better to disable the test.

**Review by rhaschke — CHANGES_REQUESTED** [review]
Thanks for your contribution. I added some more commits to your branch to fix issues I observed when testing this. Please check!
I'm not an expert in mongodb. How can we provide a .test file to eventually enable the unit test again?
What's needed to (setup and) start the mongodb server?

File: test/test_mongo_helpers.h (hunk 1 above)
Comment (christian-rauch) [reply test/test_mongo_helpers.h]: `downcastMetadata` is not a class method (as far as I observed in the code base), i.e. the `const` qualifier does not apply to any members. In fact, my compiler (gcc 7.3.0 on Ubuntu 18.04) is complaining about this.

File: CMakeLists.txt (hunk 3 above)
Comment (christian-rauch) [reply CMakeLists.txt]: The problem is the instance of the ROS master, which is needed for e.g. the time inside the test. I am not sure how this can be started in a test. Can you point me to some documentation?
I basically disabled it because the tests were dysfunctional for a long time now.

File: test/test_mongo_helpers.h (hunk 1 above)
Comment (rhaschke) [reply test/test_mongo_helpers.h]: I see.

File: CMakeLists.txt (hunk 3 above)
Comment (rhaschke) [reply CMakeLists.txt]: That's only the first issue. Running a roscore locally, the next issue is the missing mongodb server.
Hence my question: How do I run this one?

Comment (christian-rauch) [reply CMakeLists.txt]: I don't know. The bulk of the CI is done in this script: https://github.com/ros-planning/moveit_ci/blob/master/travis.sh. I assume The remaining services (ROS, mongoDB) need to be started in the docker container.
@davetcoleman Can you add an instance of the ROS master and the mongoDB server to the docker container?

Comment (rhaschke) [reply CMakeLists.txt]: You misunderstood: I know about the travis.sh script. I also know how to have a ros master for tests.
I guess, you manually verified the tests on your locally machine, running a mongodb server there.
How did you start and configure your mongodb server?

Comment (christian-rauch) [reply CMakeLists.txt]: Ah yes, sorry. I installed the meta package `mongodb` on Ubuntu via apt. I think the server started automatically after this with the correct port. There was no need for special configuration. The database is actually stored locally in a folder (given by launch file argument `moveit_warehouse_database_path`).

**Comment by christian-rauch** [comment]
I am happy to split this PR into two: one for fixing the original issue (null pointer exception when connecting to DB) and one for fixing the CI tests.
I am mostly interested in fixing the original issue since it prevents me from using `warehouse_ros_mongo`. However, the tests are failing for a long time now and they should be fixed or removed.

File: CMakeLists.txt (hunk 3 above)
Comment (rhaschke) [reply CMakeLists.txt]: Which launch file you talk about? Can you provide a pointer?
mongodb should be automatically installed by rosdep inside the docker container. This shouldn't be the issue.

**Comment by rhaschke** [comment]
OK. Giving up too. Travis hangs when running the test. I don't think it's related to roscore anymore.
Also, database connection attempt should timeout after 60s. Travis times out after 10min.
Merged as 68fc7fc33a1ce6662a4d0607ccbd0d04ed57b18e after cleanup of branches.

**Comment by christian-rauch** [comment]
Since commit 8345e7a5ed25a81230c90b0e9d542799974df7f5 I am getting compilation errors again:
```
/usr/include/mongo/bson/util/builder.h:427:15: error: ‘verify’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
         verify(z < maxSize);
         ~~~~~~^~~~~~~~~~~~~
/usr/include/mongo/bson/util/builder.h:426:15: note: ‘verify’ declared here, later in the translation unit
         verify(z >= 0);
         ~~~~~~^~~~~~~~
```

It was building fine on `kinetic` (the build server) and `melodic` (my local machine) before this commit.

**Comment by v4hn** [comment]
UPDATE: irrelevant, this was fixed in https://github.com/ros-planning/warehouse_ros_mongo/pull/18

@christian-rauch do you still see this problem?
If yes, please file a new issue or provide a patch to some repository to fix it.
At first glance it looks like rearranging the headers explosed a missing include in one of them.
